File: frontend/src/types/account.ts
export const Currency = {
  EUR: 'EUR',
  USD: 'USD',
  RUB: 'RUB',
  JPY: 'JPY',
} as const;

export type Currency = typeof Currency[keyof typeof Currency];

export interface Account {
  id: number;
  name: string;
  currency: Currency;
  balance: number;
  balance_in_default_currency?: number | null;
  default_currency?: Currency | null;
  exchange_rate_used?: number | null;
  description?: string | null;
  created_at: string;
  updated_at: string;
}

export interface AccountCreate {
  name: string;
  currency: Currency;
  balance: number;
  description?: string | null;
}

export interface AccountUpdate {
  name?: string;
  currency?: Currency;
  balance?: number;
  description?: string | null;
}

export interface AccountListResponse {
  items: Account[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface PaginationParams {
  skip?: number;
  limit?: number;
}

// Settings types (moved here due to Vite issue)
export interface UserSettings {
  id: number;
  default_currency: Currency;
  created_at: string;
  updated_at: string;
}

export interface UserSettingsUpdate {
  default_currency?: Currency;
}

export interface TotalBalance {
  total_balance: number;
  default_currency: Currency;
  currency_breakdown: Record<string, number>;
}

File: backend/src/jenmoney/schemas/account.py
from datetime import datetime
from enum import Enum
from typing import Annotated

from pydantic import Field

from jenmoney.schemas.base import BaseAPIModel


class Currency(str, Enum):
    EUR = "EUR"
    USD = "USD"
    RUB = "RUB"
    JPY = "JPY"


class AccountBase(BaseAPIModel):
    """Base account model with core fields that are always present."""

    name: Annotated[str, Field(min_length=1, max_length=100)]
    currency: Currency = Currency.EUR


class AccountCreate(AccountBase):
    """Model for creating a new account. All fields from base are required."""

    balance: Annotated[float, Field(ge=0)] = 0.0
    description: str | None = None


class AccountUpdate(BaseAPIModel):
    """Model for updating an account. All fields are optional."""

    name: Annotated[str, Field(min_length=1, max_length=100)] | None = None
    currency: Currency | None = None
    balance: Annotated[float, Field(ge=0)] | None = None
    description: str | None = None


class AccountResponse(AccountBase):
    """Model for account responses. Includes all fields plus metadata."""

    id: int
    balance: float
    percentage_of_total: float | None = None
    balance_in_default_currency: float | None = None
    default_currency: Currency | None = None
    exchange_rate_used: float | None = None
    description: str | None = None
    created_at: datetime
    updated_at: datetime


class AccountInDB(AccountResponse):
    """Internal model with all database fields."""

    pass


class AccountListResponse(BaseAPIModel):
    """Paginated response for account listings."""

    items: list[AccountResponse]
    total: int
    page: int = 1
    size: int = 10
    pages: int


class TotalBalanceResponse(BaseAPIModel):
    """Response for total balance calculation."""

    total_balance: float
    default_currency: Currency
    currency_breakdown: dict[str, float]
